// Code generated by ent, DO NOT EDIT.

package ent

import (
	"context"
	"task-svc/ent/predicate"
	"task-svc/ent/subtask"

	"entgo.io/ent/dialect/sql"
	"entgo.io/ent/dialect/sql/sqlgraph"
	"entgo.io/ent/schema/field"
)

// SubTaskDelete is the builder for deleting a SubTask entity.
type SubTaskDelete struct {
	config
	hooks    []Hook
	mutation *SubTaskMutation
}

// Where appends a list predicates to the SubTaskDelete builder.
func (std *SubTaskDelete) Where(ps ...predicate.SubTask) *SubTaskDelete {
	std.mutation.Where(ps...)
	return std
}

// Exec executes the deletion query and returns how many vertices were deleted.
func (std *SubTaskDelete) Exec(ctx context.Context) (int, error) {
	return withHooks(ctx, std.sqlExec, std.mutation, std.hooks)
}

// ExecX is like Exec, but panics if an error occurs.
func (std *SubTaskDelete) ExecX(ctx context.Context) int {
	n, err := std.Exec(ctx)
	if err != nil {
		panic(err)
	}
	return n
}

func (std *SubTaskDelete) sqlExec(ctx context.Context) (int, error) {
	_spec := sqlgraph.NewDeleteSpec(subtask.Table, sqlgraph.NewFieldSpec(subtask.FieldID, field.TypeUUID))
	if ps := std.mutation.predicates; len(ps) > 0 {
		_spec.Predicate = func(selector *sql.Selector) {
			for i := range ps {
				ps[i](selector)
			}
		}
	}
	affected, err := sqlgraph.DeleteNodes(ctx, std.driver, _spec)
	if err != nil && sqlgraph.IsConstraintError(err) {
		err = &ConstraintError{msg: err.Error(), wrap: err}
	}
	std.mutation.done = true
	return affected, err
}

// SubTaskDeleteOne is the builder for deleting a single SubTask entity.
type SubTaskDeleteOne struct {
	std *SubTaskDelete
}

// Where appends a list predicates to the SubTaskDelete builder.
func (stdo *SubTaskDeleteOne) Where(ps ...predicate.SubTask) *SubTaskDeleteOne {
	stdo.std.mutation.Where(ps...)
	return stdo
}

// Exec executes the deletion query.
func (stdo *SubTaskDeleteOne) Exec(ctx context.Context) error {
	n, err := stdo.std.Exec(ctx)
	switch {
	case err != nil:
		return err
	case n == 0:
		return &NotFoundError{subtask.Label}
	default:
		return nil
	}
}

// ExecX is like Exec, but panics if an error occurs.
func (stdo *SubTaskDeleteOne) ExecX(ctx context.Context) {
	if err := stdo.Exec(ctx); err != nil {
		panic(err)
	}
}
